{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "import matplotlib\n",
    "matplotlib.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "import pylab\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "import gzip\n",
    "import itertools\n",
    "import operator\n",
    "import subprocess\n",
    "import twobitreader\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import pysam\n",
    "\n",
    "from uditas.uditas_replace_helpers import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  Pipeline1\n",
    "# This, for the most part, is the UDITAS pipeline modified to allow for REPLACE targeting. \n",
    "\n",
    "## There is a second pipeline below to do global alignments to quantify large deletions, translocatins, concatomers, etc by aligning the sequence past the break site to the entire genome+targeting sequence.\n",
    "\n",
    "## Inputs\n",
    "#### next few window are different inputs to make it run well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#things taken from original uditas args\n",
    "\n",
    "check_plasmid_insertions = 1\n",
    "ncpu = 4\n",
    "window_size = 15\n",
    "amplicon_window_around_cut = 1000\n",
    "min_MAPQ = 5\n",
    "min_AS = -180\n",
    "process_AMP_seq_run = 0 #off"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/P_Eric4_Tn5/10000Reads\n"
     ]
    }
   ],
   "source": [
    "#Directory This is the same for all files from the same seq run\n",
    "\n",
    "#this is a minimal directory of only ~300 files for quick debugging\n",
    "directory = '/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/P_Eric4_Tn5/10000Reads'\n",
    "\n",
    "#directory = '/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/P_Eric4_Tn5'\n",
    "\n",
    "print(directory)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### make sure the following primer and downstream seq is unique and for the targeting exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "this is for polb fwd (row 0 in sheet)\n"
     ]
    }
   ],
   "source": [
    "#primer sequences for mispriming. Primer sequence + 12nt downstream \n",
    "#this will only be avialable for read2\n",
    "\n",
    "##### the sequences need to be capital for this to work########\n",
    "\n",
    "#polb fwd \n",
    "print('this is for polb fwd (row 0 in sheet)')\n",
    "primer_seq_plus_downstream ='TGGTTTTATTTCACCCCATGATAGTAATTATATCACT'\n",
    "\n",
    "\n",
    "#print('this is for pipeline1 polb rev (row 1 in sheet)')\n",
    "#primer_seq_plus_downstream ='ACAAAAGAGGCCAAGCTGGAGCAGGAAATAGATGC'\n",
    "\n",
    "\n",
    "#print('this is for pipeline1 mCherryRev, row2 in sheet')\n",
    "#make sure letters are Capital! (row 2 in sheet)\n",
    "#primer_seq_plus_downstream ='GTTAGCAGACTTCCTCTGCCCTCAATCTTTAAGAA'\n",
    "\n",
    "\n",
    "#print('this is for mCherryFWD, row3 in sheet')\n",
    "#primer_seq_plus_downstream ='GCAGGAGCTCGTCGACCCATGGGGGCCCGCCCC'\n",
    "\n",
    "\n",
    "#check what is printed below!!!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sheets information input (change for each targeting exp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NGS_req-ID                                                                     A000H2GWGF\n",
       "name                                                                        Tn5_Pol50_pbF\n",
       "Sample                                                                               Polb\n",
       "description                                   MiniSeq_K562_PolbHIROS_50per_unsorted_polbF\n",
       "Control sample (Y/N)                                                                    N\n",
       "Notes                                                                                 NaN\n",
       "Dilution                                                                              NaN\n",
       "Cell name_type                                                                       K562\n",
       "UMI_Len                                                                        NNNNNNNNNN\n",
       "IndexI7Primer                                                                      prE368\n",
       "I7_Index_ID                                                          P7_N701_SBS12nextera\n",
       "index_I1                                                                             N701\n",
       "barcode_I1                                                                       TAAGGCGA\n",
       "I5_Index_ID                                                              P5_UMI_N50_Tn5-A\n",
       "index_I2                                                                             N501\n",
       "barcode_I2                                                                       GCGATCTA\n",
       "primer1_ForNest                                                                     prE86\n",
       "primer                                                                             prE361\n",
       "genome                                                                               hg38\n",
       "chr                                                                                  chr8\n",
       "start                                                                         4.23497e+07\n",
       "end                                                                           4.23498e+07\n",
       "strand                                                                                  +\n",
       "guide_1                                                                            Polb#3\n",
       "sequence_guide_1                                                     TAATTTTGTGTGGGTCACCC\n",
       "genome_guide_1                                                                       hg38\n",
       "chr_guide_1                                                                          chr8\n",
       "start_guide_1                                                                    42349873\n",
       "end_guide_1                                                                      42349892\n",
       "strand_guide_1                                                                          +\n",
       "guide_2                                                                            Polb#4\n",
       "sequence_guide_2                                                     TGAAACCAGTTTGGTTACCC\n",
       "genome_guide_2                                                                       hg38\n",
       "chr_guide_2                                                                          chr8\n",
       "start_guide_2                                                                    42350280\n",
       "end_guide_2                                                                      42350299\n",
       "strand_guide_2                                                                          +\n",
       "guide_3                                                                               NaN\n",
       "sequence_guide_3                                                                      NaN\n",
       "genome_guide_3                                                                        NaN\n",
       "chr_guide_3                                                                           NaN\n",
       "start_guide_3                                                                         NaN\n",
       "end_guide_3                                                                           NaN\n",
       "strand_guide_3                                                                        NaN\n",
       "genome_plus_targeting                                              hg38_plus_targetvector\n",
       "Replace_Doner_Name                                                               pE038_MC\n",
       "Replace_Donor                           TGACCCACACAAAATTAGTCTTTTAGCAGACTGGTATGTTTCCAAT...\n",
       "Donor (single strand; double strand)                                                  NaN\n",
       "plamid_name                                                            pE049_WO_redundant\n",
       "plasmid_sequence                        ttacataacttacggtaaatggcccgcctggctgaccgcccaacga...\n",
       "genomic_background                                                                    NaN\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Function to make the 'amplicon_info' list. Taking the line of our experiment csv file\n",
    "\n",
    "def get_csv_data(dir_sample, line_of_data_from_csv):\n",
    "    sample_info_filename = os.path.join(dir_sample, 'sample_info.csv')\n",
    "    experiments = pd.read_csv(sample_info_filename)\n",
    "    return experiments.loc[line_of_data_from_csv]\n",
    "\n",
    "\n",
    "#the amplicon info is related to the line on the csv file. It is indexed from 0. For PolbF we use 0.\n",
    "#it is on my extera space so need to make sure that is mounted. (first thing to check if it throws and error)\n",
    "\n",
    "#0 is because this polbFwd is in the first line. Need to change.\n",
    "amplicon_info = get_csv_data(directory, 0)\n",
    "\n",
    "#1 is because this polbRev \n",
    "#amplicon_info = get_csv_data(directory, 1)\n",
    "\n",
    "#2 is because this mcherry_rev\n",
    "#amplicon_info = get_csv_data(directory, 2)\n",
    "\n",
    "#3 is because this mcherry_fwd\n",
    "#amplicon_info = get_csv_data(directory, 3)\n",
    "\n",
    "\n",
    "###########check waht is printed below #######\n",
    "amplicon_info"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set global Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes/hg38.2bit\n",
      "env: BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes\n"
     ]
    }
   ],
   "source": [
    "# Assign the file_genome_2bit location. This is needed for pulling sequence from the referene genome by location\n",
    "assembly = amplicon_info['genome']\n",
    "file_genome_2bit = os.path.join('/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes', assembly + '.2bit')\n",
    "print(file_genome_2bit)\n",
    "\n",
    "# BOWTIE2_INDEXES are needed for global alignments\n",
    "#not sure if this will work\n",
    "#normally in bash: export BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes\n",
    "#check in bash: > ECHO $GENOMES_2BIT\n",
    "\n",
    "#this is the refrence genome for Bowtie from Illumina. It does not have our targeting vector\n",
    "%env BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes\n",
    "\n",
    "#check the env variables\n",
    "#%env"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions to process the reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#I imported them from my helper file to keep this space cleaner"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Discard Mispriming Reads\n",
    "When you put a universal primer on the ends of everything, every mispriming event will amplify. An effect we normally don't deal with. I did nested PCR to reduce this. However 85% of the alignments in the UDITAS data I looked at seemed to be mispriming. They did all their blasting and analysis before removing mispriming. But to save computational power and remove error early on I will discard mispriming events. \n",
    "They discard these only for plasmid alignments analyze_alignments_plasmid for some reason which comes from the bam file.\n",
    "\n",
    "By eye it looks like 50-90% of my reads are correctly primed which is amaizng. Nesting helepd a lot.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reads_in_experiment_list_count 323\n",
      "reads_not_in_experiment_list_count 30\n"
     ]
    }
   ],
   "source": [
    "correct_priming(directory, amplicon_info, primer_seq_plus_downstream)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Trim off illumina adapter from shorter reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Running the trimming\n",
    "\n",
    "trim_fastq(directory, amplicon_info, 0)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Running the reference plasmid creation\n",
    "create_plasmid_reference(directory, amplicon_info)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'replace'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Test the get_reaction_type is working\n",
    "#uditas_replace_helpers.get_reaction_type(amplicon_info)\n",
    "get_reaction_type(amplicon_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make the amplicons by calling the create_amplicon function\n",
    "\n",
    "create_amplicon(directory, amplicon_info, file_genome_2bit)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "#try out the alignment to the plasmid\n",
    "align_plasmid_local(directory, amplicon_info, ncpu=4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "#extract the unmapped reads\n",
    "extract_unmapped_reads_plasmid(directory, amplicon_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_plus_plasmid_total_reads</th>\n",
       "      <th>target_plus_plasmid_total_reads_collapsed</th>\n",
       "      <th>plasmid_only_total_reads</th>\n",
       "      <th>plasmid_only_total_reads_collapsed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   target_plus_plasmid_total_reads  target_plus_plasmid_total_reads_collapsed  \\\n",
       "0                                0                                          0   \n",
       "\n",
       "   plasmid_only_total_reads  plasmid_only_total_reads_collapsed  \n",
       "0                         0                                   0  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#run the plasmid analysis to coudn plasmid integration events\n",
    "result_plasmid_df = analyze_alignments_plasmid(directory, amplicon_info, min_MAPQ, file_genome_2bit, True)\n",
    "result_plasmid_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#align against our suite of amplicons\n",
    "# took about 5 min with 300k reads\n",
    "align_amplicon(directory, amplicon_info, check_plasmid_insertions, ncpu)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#this will extract unmap reads new folder (files that did not align to the predicted structural variants)\n",
    "extract_unmapped_reads_amplicons(directory, amplicon_info)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## This section is to do the analysis of the amplicon alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of reads before mispriming cleanup 353\n",
      "number of unique umis before mispriming cleanup 160\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>all_amplicons_total_reads</th>\n",
       "      <th>all_amplicons_total_reads_collapsed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>226</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   all_amplicons_total_reads  all_amplicons_total_reads_collapsed\n",
       "0                        226                                   86"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_reads_in_all_amplicons_df = analyze_alignments_all_amplicons(directory, amplicon_info, min_MAPQ, min_AS)\n",
    "result_reads_in_all_amplicons_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt_cut1_total_reads</th>\n",
       "      <th>wt_cut1_total_indels</th>\n",
       "      <th>wt_cut1_total_deletions</th>\n",
       "      <th>wt_cut1_total_insertions</th>\n",
       "      <th>wt_cut1_total_reads_collapsed</th>\n",
       "      <th>wt_cut1_total_indels_collapsed</th>\n",
       "      <th>wt_cut1_total_deletions_collapsed</th>\n",
       "      <th>wt_cut1_total_insertions_collapsed</th>\n",
       "      <th>wt_cut2_total_reads</th>\n",
       "      <th>wt_cut2_total_indels</th>\n",
       "      <th>...</th>\n",
       "      <th>1a_1a_cut1_total_insertions_collapsed</th>\n",
       "      <th>2b_2b_cut1_total_reads</th>\n",
       "      <th>2b_2b_cut1_total_indels</th>\n",
       "      <th>2b_2b_cut1_total_deletions</th>\n",
       "      <th>2b_2b_cut1_total_insertions</th>\n",
       "      <th>2b_2b_cut1_total_reads_collapsed</th>\n",
       "      <th>2b_2b_cut1_total_indels_collapsed</th>\n",
       "      <th>2b_2b_cut1_total_deletions_collapsed</th>\n",
       "      <th>2b_2b_cut1_total_insertions_collapsed</th>\n",
       "      <th>median_fragment_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>37</td>\n",
       "      <td>21</td>\n",
       "      <td>6</td>\n",
       "      <td>15</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>208.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 113 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   wt_cut1_total_reads  wt_cut1_total_indels  wt_cut1_total_deletions  \\\n",
       "0                   37                    21                        6   \n",
       "\n",
       "   wt_cut1_total_insertions  wt_cut1_total_reads_collapsed  \\\n",
       "0                        15                             12   \n",
       "\n",
       "   wt_cut1_total_indels_collapsed  wt_cut1_total_deletions_collapsed  \\\n",
       "0                               6                                  2   \n",
       "\n",
       "   wt_cut1_total_insertions_collapsed  wt_cut2_total_reads  \\\n",
       "0                                   4                    0   \n",
       "\n",
       "   wt_cut2_total_indels  ...  1a_1a_cut1_total_insertions_collapsed  \\\n",
       "0                     0  ...                                      0   \n",
       "\n",
       "   2b_2b_cut1_total_reads  2b_2b_cut1_total_indels  \\\n",
       "0                       0                        0   \n",
       "\n",
       "   2b_2b_cut1_total_deletions  2b_2b_cut1_total_insertions  \\\n",
       "0                           0                            0   \n",
       "\n",
       "   2b_2b_cut1_total_reads_collapsed  2b_2b_cut1_total_indels_collapsed  \\\n",
       "0                                 0                                  0   \n",
       "\n",
       "   2b_2b_cut1_total_deletions_collapsed  \\\n",
       "0                                     0   \n",
       "\n",
       "   2b_2b_cut1_total_insertions_collapsed  median_fragment_size  \n",
       "0                                      0                 208.0  \n",
       "\n",
       "[1 rows x 113 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#took \n",
    "result_amplicon_df = analyze_alignments(directory, amplicon_info, window_size, amplicon_window_around_cut, min_MAPQ, min_AS)\n",
    "result_amplicon_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### This is the Global alignment of the remaining reads\n",
    "\n",
    "#### What I need here is to modify it so that "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run the allignment\n",
    "assembly = amplicon_info['genome']\n",
    "align_genome_global(directory, amplicon_info, assembly)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the number of alignments that were attempted to align 0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genomewide_total_reads</th>\n",
       "      <th>genomewide_total_reads_collapsed</th>\n",
       "      <th>genomewide_target_only_reads</th>\n",
       "      <th>genomewide_target_only_reads_collapsed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   genomewide_total_reads  genomewide_total_reads_collapsed  \\\n",
       "0                       0                                 0   \n",
       "\n",
       "   genomewide_target_only_reads  genomewide_target_only_reads_collapsed  \n",
       "0                             0                                       0  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Do the global alignment analysis\n",
    "results_genome_global_df = analyze_alignments_genome_global(directory, amplicon_info, min_MAPQ, min_AS,  file_genome_2bit)\n",
    "results_genome_global_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summerize data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "read_count = pd.DataFrame()\n",
    "read_count.loc[0,'read_count'] = count_reads(directory, amplicon_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_count</th>\n",
       "      <th>target_plus_plasmid_total_reads</th>\n",
       "      <th>target_plus_plasmid_total_reads_collapsed</th>\n",
       "      <th>plasmid_only_total_reads</th>\n",
       "      <th>plasmid_only_total_reads_collapsed</th>\n",
       "      <th>all_amplicons_total_reads</th>\n",
       "      <th>all_amplicons_total_reads_collapsed</th>\n",
       "      <th>genomewide_total_reads</th>\n",
       "      <th>genomewide_total_reads_collapsed</th>\n",
       "      <th>genomewide_target_only_reads</th>\n",
       "      <th>genomewide_target_only_reads_collapsed</th>\n",
       "      <th>total_aligned</th>\n",
       "      <th>total_aligned_collapsed</th>\n",
       "      <th>percent_aligned</th>\n",
       "      <th>percent_aligned_all_amplicons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>323.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>226</td>\n",
       "      <td>86</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>226</td>\n",
       "      <td>86</td>\n",
       "      <td>69.96904</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   read_count  target_plus_plasmid_total_reads  \\\n",
       "0       323.0                                0   \n",
       "\n",
       "   target_plus_plasmid_total_reads_collapsed  plasmid_only_total_reads  \\\n",
       "0                                          0                         0   \n",
       "\n",
       "   plasmid_only_total_reads_collapsed  all_amplicons_total_reads  \\\n",
       "0                                   0                        226   \n",
       "\n",
       "   all_amplicons_total_reads_collapsed  genomewide_total_reads  \\\n",
       "0                                   86                       0   \n",
       "\n",
       "   genomewide_total_reads_collapsed  genomewide_target_only_reads  \\\n",
       "0                                 0                             0   \n",
       "\n",
       "   genomewide_target_only_reads_collapsed  total_aligned  \\\n",
       "0                                       0            226   \n",
       "\n",
       "   total_aligned_collapsed  percent_aligned  percent_aligned_all_amplicons  \n",
       "0                       86         69.96904                          100.0  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#summerize the data\n",
    "\n",
    "summary_all_alignments = get_summary_all_alignments(directory, amplicon_info, read_count.loc[[0]], result_plasmid_df, result_reads_in_all_amplicons_df, results_genome_global_df)\n",
    "summary_all_alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NGS_req-ID</th>\n",
       "      <th>name</th>\n",
       "      <th>Sample</th>\n",
       "      <th>description</th>\n",
       "      <th>Control sample (Y/N)</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Dilution</th>\n",
       "      <th>Cell name_type</th>\n",
       "      <th>UMI_Len</th>\n",
       "      <th>IndexI7Primer</th>\n",
       "      <th>...</th>\n",
       "      <th>doner_tail_tail_cut1_total_reads_percent</th>\n",
       "      <th>doner_head_tail_cut1_total_reads_percent</th>\n",
       "      <th>doner_head_head_cut1_total_reads_percent</th>\n",
       "      <th>1a_1a_cut1_total_reads_percent</th>\n",
       "      <th>2b_2b_cut1_total_reads_percent</th>\n",
       "      <th>target_plus_plasmid_total_reads_percent</th>\n",
       "      <th>plasmid_only_total_reads_percent</th>\n",
       "      <th>total_aligned_junctions_collapsed</th>\n",
       "      <th>Type</th>\n",
       "      <th>Percent Editing</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbF</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE368</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.111111</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>wt_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>17.647059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbR</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE369</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_mCh</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_bpa</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE372</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbF</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE368</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.111111</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>wt_cut2_total_reads_collapsed_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbR</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE369</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_mCh</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_bpa</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE372</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbF</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE368</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.111111</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>large_deletion_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>35.294118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbR</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE369</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_mCh</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_bpa</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE372</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbF</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE368</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.111111</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>large_inversion_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>16.176471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbR</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE369</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_mCh</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_bpa</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE372</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbF</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE368</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.111111</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>large_inversion_cut2_total_reads_collapsed_per...</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbR</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE369</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut2_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_mCh</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut2_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_bpa</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut2_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE372</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut2_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut2_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbF</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE368</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.111111</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>doner_head_head_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbR</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE369</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>doner_head_head_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_mCh</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>doner_head_head_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_bpa</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>doner_head_head_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE372</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>doner_head_head_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>doner_head_head_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbF</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE368</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.111111</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>1a_1a_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>1.470588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbR</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE369</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_mCh</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_bpa</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE372</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbF</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE368</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.111111</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>2b_2b_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbR</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE369</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_mCh</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_bpa</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE372</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbF</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE368</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.111111</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>target_plus_plasmid_total_reads_collapsed_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbR</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE369</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_mCh</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_bpa</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE372</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbF</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE368</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.111111</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>plasmid_only_total_reads_collapsed_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbR</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE369</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_mCh</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_bpa</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE372</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>96 rows Ã— 187 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    NGS_req-ID           name Sample  \\\n",
       "0   A000H2GWGF  Tn5_Pol50_pbF   Polb   \n",
       "1   A000H2GWGF  Tn5_Pol50_pbR   Polb   \n",
       "2   A000H2GWGF  Tn5_Pol50_mCh   Polb   \n",
       "3   A000H2GWGF  Tn5_Pol50_bpa   Polb   \n",
       "4   A000H2GWGF     Tn5_WT_pbF   Polb   \n",
       "5   A000H2GWGF     Tn5_WT_pbR   Polb   \n",
       "6   A000H2GWGF  Tn5_Pol50_pbF   Polb   \n",
       "7   A000H2GWGF  Tn5_Pol50_pbR   Polb   \n",
       "8   A000H2GWGF  Tn5_Pol50_mCh   Polb   \n",
       "9   A000H2GWGF  Tn5_Pol50_bpa   Polb   \n",
       "10  A000H2GWGF     Tn5_WT_pbF   Polb   \n",
       "11  A000H2GWGF     Tn5_WT_pbR   Polb   \n",
       "12  A000H2GWGF  Tn5_Pol50_pbF   Polb   \n",
       "13  A000H2GWGF  Tn5_Pol50_pbR   Polb   \n",
       "14  A000H2GWGF  Tn5_Pol50_mCh   Polb   \n",
       "15  A000H2GWGF  Tn5_Pol50_bpa   Polb   \n",
       "16  A000H2GWGF     Tn5_WT_pbF   Polb   \n",
       "17  A000H2GWGF     Tn5_WT_pbR   Polb   \n",
       "18  A000H2GWGF  Tn5_Pol50_pbF   Polb   \n",
       "19  A000H2GWGF  Tn5_Pol50_pbR   Polb   \n",
       "20  A000H2GWGF  Tn5_Pol50_mCh   Polb   \n",
       "21  A000H2GWGF  Tn5_Pol50_bpa   Polb   \n",
       "22  A000H2GWGF     Tn5_WT_pbF   Polb   \n",
       "23  A000H2GWGF     Tn5_WT_pbR   Polb   \n",
       "24  A000H2GWGF  Tn5_Pol50_pbF   Polb   \n",
       "25  A000H2GWGF  Tn5_Pol50_pbR   Polb   \n",
       "26  A000H2GWGF  Tn5_Pol50_mCh   Polb   \n",
       "27  A000H2GWGF  Tn5_Pol50_bpa   Polb   \n",
       "28  A000H2GWGF     Tn5_WT_pbF   Polb   \n",
       "29  A000H2GWGF     Tn5_WT_pbR   Polb   \n",
       "..         ...            ...    ...   \n",
       "66  A000H2GWGF  Tn5_Pol50_pbF   Polb   \n",
       "67  A000H2GWGF  Tn5_Pol50_pbR   Polb   \n",
       "68  A000H2GWGF  Tn5_Pol50_mCh   Polb   \n",
       "69  A000H2GWGF  Tn5_Pol50_bpa   Polb   \n",
       "70  A000H2GWGF     Tn5_WT_pbF   Polb   \n",
       "71  A000H2GWGF     Tn5_WT_pbR   Polb   \n",
       "72  A000H2GWGF  Tn5_Pol50_pbF   Polb   \n",
       "73  A000H2GWGF  Tn5_Pol50_pbR   Polb   \n",
       "74  A000H2GWGF  Tn5_Pol50_mCh   Polb   \n",
       "75  A000H2GWGF  Tn5_Pol50_bpa   Polb   \n",
       "76  A000H2GWGF     Tn5_WT_pbF   Polb   \n",
       "77  A000H2GWGF     Tn5_WT_pbR   Polb   \n",
       "78  A000H2GWGF  Tn5_Pol50_pbF   Polb   \n",
       "79  A000H2GWGF  Tn5_Pol50_pbR   Polb   \n",
       "80  A000H2GWGF  Tn5_Pol50_mCh   Polb   \n",
       "81  A000H2GWGF  Tn5_Pol50_bpa   Polb   \n",
       "82  A000H2GWGF     Tn5_WT_pbF   Polb   \n",
       "83  A000H2GWGF     Tn5_WT_pbR   Polb   \n",
       "84  A000H2GWGF  Tn5_Pol50_pbF   Polb   \n",
       "85  A000H2GWGF  Tn5_Pol50_pbR   Polb   \n",
       "86  A000H2GWGF  Tn5_Pol50_mCh   Polb   \n",
       "87  A000H2GWGF  Tn5_Pol50_bpa   Polb   \n",
       "88  A000H2GWGF     Tn5_WT_pbF   Polb   \n",
       "89  A000H2GWGF     Tn5_WT_pbR   Polb   \n",
       "90  A000H2GWGF  Tn5_Pol50_pbF   Polb   \n",
       "91  A000H2GWGF  Tn5_Pol50_pbR   Polb   \n",
       "92  A000H2GWGF  Tn5_Pol50_mCh   Polb   \n",
       "93  A000H2GWGF  Tn5_Pol50_bpa   Polb   \n",
       "94  A000H2GWGF     Tn5_WT_pbF   Polb   \n",
       "95  A000H2GWGF     Tn5_WT_pbR   Polb   \n",
       "\n",
       "                                          description Control sample (Y/N)  \\\n",
       "0         MiniSeq_K562_PolbHIROS_50per_unsorted_polbF                    N   \n",
       "1         MiniSeq_K562_PolbHIROS_50per_unsorted_polbR                    N   \n",
       "2   MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...                    N   \n",
       "3   MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd                    N   \n",
       "4   MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "5   MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "6         MiniSeq_K562_PolbHIROS_50per_unsorted_polbF                    N   \n",
       "7         MiniSeq_K562_PolbHIROS_50per_unsorted_polbR                    N   \n",
       "8   MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...                    N   \n",
       "9   MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd                    N   \n",
       "10  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "11  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "12        MiniSeq_K562_PolbHIROS_50per_unsorted_polbF                    N   \n",
       "13        MiniSeq_K562_PolbHIROS_50per_unsorted_polbR                    N   \n",
       "14  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...                    N   \n",
       "15  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd                    N   \n",
       "16  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "17  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "18        MiniSeq_K562_PolbHIROS_50per_unsorted_polbF                    N   \n",
       "19        MiniSeq_K562_PolbHIROS_50per_unsorted_polbR                    N   \n",
       "20  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...                    N   \n",
       "21  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd                    N   \n",
       "22  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "23  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "24        MiniSeq_K562_PolbHIROS_50per_unsorted_polbF                    N   \n",
       "25        MiniSeq_K562_PolbHIROS_50per_unsorted_polbR                    N   \n",
       "26  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...                    N   \n",
       "27  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd                    N   \n",
       "28  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "29  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "..                                                ...                  ...   \n",
       "66        MiniSeq_K562_PolbHIROS_50per_unsorted_polbF                    N   \n",
       "67        MiniSeq_K562_PolbHIROS_50per_unsorted_polbR                    N   \n",
       "68  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...                    N   \n",
       "69  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd                    N   \n",
       "70  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "71  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "72        MiniSeq_K562_PolbHIROS_50per_unsorted_polbF                    N   \n",
       "73        MiniSeq_K562_PolbHIROS_50per_unsorted_polbR                    N   \n",
       "74  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...                    N   \n",
       "75  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd                    N   \n",
       "76  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "77  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "78        MiniSeq_K562_PolbHIROS_50per_unsorted_polbF                    N   \n",
       "79        MiniSeq_K562_PolbHIROS_50per_unsorted_polbR                    N   \n",
       "80  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...                    N   \n",
       "81  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd                    N   \n",
       "82  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "83  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "84        MiniSeq_K562_PolbHIROS_50per_unsorted_polbF                    N   \n",
       "85        MiniSeq_K562_PolbHIROS_50per_unsorted_polbR                    N   \n",
       "86  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...                    N   \n",
       "87  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd                    N   \n",
       "88  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "89  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "90        MiniSeq_K562_PolbHIROS_50per_unsorted_polbF                    N   \n",
       "91        MiniSeq_K562_PolbHIROS_50per_unsorted_polbR                    N   \n",
       "92  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...                    N   \n",
       "93  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd                    N   \n",
       "94  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "95  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "\n",
       "    Notes  Dilution Cell name_type     UMI_Len IndexI7Primer  ...  \\\n",
       "0     NaN       NaN           K562  NNNNNNNNNN        prE368  ...   \n",
       "1     NaN       NaN           K562  NNNNNNNNNN        prE369  ...   \n",
       "2     NaN       NaN           K562  NNNNNNNNNN        prE370  ...   \n",
       "3     NaN       NaN           K562  NNNNNNNNNN        prE371  ...   \n",
       "4     NaN       NaN           K562  NNNNNNNNNN        prE372  ...   \n",
       "5     NaN       NaN           K562  NNNNNNNNNN        prE373  ...   \n",
       "6     NaN       NaN           K562  NNNNNNNNNN        prE368  ...   \n",
       "7     NaN       NaN           K562  NNNNNNNNNN        prE369  ...   \n",
       "8     NaN       NaN           K562  NNNNNNNNNN        prE370  ...   \n",
       "9     NaN       NaN           K562  NNNNNNNNNN        prE371  ...   \n",
       "10    NaN       NaN           K562  NNNNNNNNNN        prE372  ...   \n",
       "11    NaN       NaN           K562  NNNNNNNNNN        prE373  ...   \n",
       "12    NaN       NaN           K562  NNNNNNNNNN        prE368  ...   \n",
       "13    NaN       NaN           K562  NNNNNNNNNN        prE369  ...   \n",
       "14    NaN       NaN           K562  NNNNNNNNNN        prE370  ...   \n",
       "15    NaN       NaN           K562  NNNNNNNNNN        prE371  ...   \n",
       "16    NaN       NaN           K562  NNNNNNNNNN        prE372  ...   \n",
       "17    NaN       NaN           K562  NNNNNNNNNN        prE373  ...   \n",
       "18    NaN       NaN           K562  NNNNNNNNNN        prE368  ...   \n",
       "19    NaN       NaN           K562  NNNNNNNNNN        prE369  ...   \n",
       "20    NaN       NaN           K562  NNNNNNNNNN        prE370  ...   \n",
       "21    NaN       NaN           K562  NNNNNNNNNN        prE371  ...   \n",
       "22    NaN       NaN           K562  NNNNNNNNNN        prE372  ...   \n",
       "23    NaN       NaN           K562  NNNNNNNNNN        prE373  ...   \n",
       "24    NaN       NaN           K562  NNNNNNNNNN        prE368  ...   \n",
       "25    NaN       NaN           K562  NNNNNNNNNN        prE369  ...   \n",
       "26    NaN       NaN           K562  NNNNNNNNNN        prE370  ...   \n",
       "27    NaN       NaN           K562  NNNNNNNNNN        prE371  ...   \n",
       "28    NaN       NaN           K562  NNNNNNNNNN        prE372  ...   \n",
       "29    NaN       NaN           K562  NNNNNNNNNN        prE373  ...   \n",
       "..    ...       ...            ...         ...           ...  ...   \n",
       "66    NaN       NaN           K562  NNNNNNNNNN        prE368  ...   \n",
       "67    NaN       NaN           K562  NNNNNNNNNN        prE369  ...   \n",
       "68    NaN       NaN           K562  NNNNNNNNNN        prE370  ...   \n",
       "69    NaN       NaN           K562  NNNNNNNNNN        prE371  ...   \n",
       "70    NaN       NaN           K562  NNNNNNNNNN        prE372  ...   \n",
       "71    NaN       NaN           K562  NNNNNNNNNN        prE373  ...   \n",
       "72    NaN       NaN           K562  NNNNNNNNNN        prE368  ...   \n",
       "73    NaN       NaN           K562  NNNNNNNNNN        prE369  ...   \n",
       "74    NaN       NaN           K562  NNNNNNNNNN        prE370  ...   \n",
       "75    NaN       NaN           K562  NNNNNNNNNN        prE371  ...   \n",
       "76    NaN       NaN           K562  NNNNNNNNNN        prE372  ...   \n",
       "77    NaN       NaN           K562  NNNNNNNNNN        prE373  ...   \n",
       "78    NaN       NaN           K562  NNNNNNNNNN        prE368  ...   \n",
       "79    NaN       NaN           K562  NNNNNNNNNN        prE369  ...   \n",
       "80    NaN       NaN           K562  NNNNNNNNNN        prE370  ...   \n",
       "81    NaN       NaN           K562  NNNNNNNNNN        prE371  ...   \n",
       "82    NaN       NaN           K562  NNNNNNNNNN        prE372  ...   \n",
       "83    NaN       NaN           K562  NNNNNNNNNN        prE373  ...   \n",
       "84    NaN       NaN           K562  NNNNNNNNNN        prE368  ...   \n",
       "85    NaN       NaN           K562  NNNNNNNNNN        prE369  ...   \n",
       "86    NaN       NaN           K562  NNNNNNNNNN        prE370  ...   \n",
       "87    NaN       NaN           K562  NNNNNNNNNN        prE371  ...   \n",
       "88    NaN       NaN           K562  NNNNNNNNNN        prE372  ...   \n",
       "89    NaN       NaN           K562  NNNNNNNNNN        prE373  ...   \n",
       "90    NaN       NaN           K562  NNNNNNNNNN        prE368  ...   \n",
       "91    NaN       NaN           K562  NNNNNNNNNN        prE369  ...   \n",
       "92    NaN       NaN           K562  NNNNNNNNNN        prE370  ...   \n",
       "93    NaN       NaN           K562  NNNNNNNNNN        prE371  ...   \n",
       "94    NaN       NaN           K562  NNNNNNNNNN        prE372  ...   \n",
       "95    NaN       NaN           K562  NNNNNNNNNN        prE373  ...   \n",
       "\n",
       "   doner_tail_tail_cut1_total_reads_percent  \\\n",
       "0                                       0.0   \n",
       "1                                       NaN   \n",
       "2                                       NaN   \n",
       "3                                       NaN   \n",
       "4                                       NaN   \n",
       "5                                       NaN   \n",
       "6                                       0.0   \n",
       "7                                       NaN   \n",
       "8                                       NaN   \n",
       "9                                       NaN   \n",
       "10                                      NaN   \n",
       "11                                      NaN   \n",
       "12                                      0.0   \n",
       "13                                      NaN   \n",
       "14                                      NaN   \n",
       "15                                      NaN   \n",
       "16                                      NaN   \n",
       "17                                      NaN   \n",
       "18                                      0.0   \n",
       "19                                      NaN   \n",
       "20                                      NaN   \n",
       "21                                      NaN   \n",
       "22                                      NaN   \n",
       "23                                      NaN   \n",
       "24                                      0.0   \n",
       "25                                      NaN   \n",
       "26                                      NaN   \n",
       "27                                      NaN   \n",
       "28                                      NaN   \n",
       "29                                      NaN   \n",
       "..                                      ...   \n",
       "66                                      0.0   \n",
       "67                                      NaN   \n",
       "68                                      NaN   \n",
       "69                                      NaN   \n",
       "70                                      NaN   \n",
       "71                                      NaN   \n",
       "72                                      0.0   \n",
       "73                                      NaN   \n",
       "74                                      NaN   \n",
       "75                                      NaN   \n",
       "76                                      NaN   \n",
       "77                                      NaN   \n",
       "78                                      0.0   \n",
       "79                                      NaN   \n",
       "80                                      NaN   \n",
       "81                                      NaN   \n",
       "82                                      NaN   \n",
       "83                                      NaN   \n",
       "84                                      0.0   \n",
       "85                                      NaN   \n",
       "86                                      NaN   \n",
       "87                                      NaN   \n",
       "88                                      NaN   \n",
       "89                                      NaN   \n",
       "90                                      0.0   \n",
       "91                                      NaN   \n",
       "92                                      NaN   \n",
       "93                                      NaN   \n",
       "94                                      NaN   \n",
       "95                                      NaN   \n",
       "\n",
       "   doner_head_tail_cut1_total_reads_percent  \\\n",
       "0                                       0.0   \n",
       "1                                       NaN   \n",
       "2                                       NaN   \n",
       "3                                       NaN   \n",
       "4                                       NaN   \n",
       "5                                       NaN   \n",
       "6                                       0.0   \n",
       "7                                       NaN   \n",
       "8                                       NaN   \n",
       "9                                       NaN   \n",
       "10                                      NaN   \n",
       "11                                      NaN   \n",
       "12                                      0.0   \n",
       "13                                      NaN   \n",
       "14                                      NaN   \n",
       "15                                      NaN   \n",
       "16                                      NaN   \n",
       "17                                      NaN   \n",
       "18                                      0.0   \n",
       "19                                      NaN   \n",
       "20                                      NaN   \n",
       "21                                      NaN   \n",
       "22                                      NaN   \n",
       "23                                      NaN   \n",
       "24                                      0.0   \n",
       "25                                      NaN   \n",
       "26                                      NaN   \n",
       "27                                      NaN   \n",
       "28                                      NaN   \n",
       "29                                      NaN   \n",
       "..                                      ...   \n",
       "66                                      0.0   \n",
       "67                                      NaN   \n",
       "68                                      NaN   \n",
       "69                                      NaN   \n",
       "70                                      NaN   \n",
       "71                                      NaN   \n",
       "72                                      0.0   \n",
       "73                                      NaN   \n",
       "74                                      NaN   \n",
       "75                                      NaN   \n",
       "76                                      NaN   \n",
       "77                                      NaN   \n",
       "78                                      0.0   \n",
       "79                                      NaN   \n",
       "80                                      NaN   \n",
       "81                                      NaN   \n",
       "82                                      NaN   \n",
       "83                                      NaN   \n",
       "84                                      0.0   \n",
       "85                                      NaN   \n",
       "86                                      NaN   \n",
       "87                                      NaN   \n",
       "88                                      NaN   \n",
       "89                                      NaN   \n",
       "90                                      0.0   \n",
       "91                                      NaN   \n",
       "92                                      NaN   \n",
       "93                                      NaN   \n",
       "94                                      NaN   \n",
       "95                                      NaN   \n",
       "\n",
       "   doner_head_head_cut1_total_reads_percent 1a_1a_cut1_total_reads_percent  \\\n",
       "0                                       0.0                       1.111111   \n",
       "1                                       NaN                            NaN   \n",
       "2                                       NaN                            NaN   \n",
       "3                                       NaN                            NaN   \n",
       "4                                       NaN                            NaN   \n",
       "5                                       NaN                            NaN   \n",
       "6                                       0.0                       1.111111   \n",
       "7                                       NaN                            NaN   \n",
       "8                                       NaN                            NaN   \n",
       "9                                       NaN                            NaN   \n",
       "10                                      NaN                            NaN   \n",
       "11                                      NaN                            NaN   \n",
       "12                                      0.0                       1.111111   \n",
       "13                                      NaN                            NaN   \n",
       "14                                      NaN                            NaN   \n",
       "15                                      NaN                            NaN   \n",
       "16                                      NaN                            NaN   \n",
       "17                                      NaN                            NaN   \n",
       "18                                      0.0                       1.111111   \n",
       "19                                      NaN                            NaN   \n",
       "20                                      NaN                            NaN   \n",
       "21                                      NaN                            NaN   \n",
       "22                                      NaN                            NaN   \n",
       "23                                      NaN                            NaN   \n",
       "24                                      0.0                       1.111111   \n",
       "25                                      NaN                            NaN   \n",
       "26                                      NaN                            NaN   \n",
       "27                                      NaN                            NaN   \n",
       "28                                      NaN                            NaN   \n",
       "29                                      NaN                            NaN   \n",
       "..                                      ...                            ...   \n",
       "66                                      0.0                       1.111111   \n",
       "67                                      NaN                            NaN   \n",
       "68                                      NaN                            NaN   \n",
       "69                                      NaN                            NaN   \n",
       "70                                      NaN                            NaN   \n",
       "71                                      NaN                            NaN   \n",
       "72                                      0.0                       1.111111   \n",
       "73                                      NaN                            NaN   \n",
       "74                                      NaN                            NaN   \n",
       "75                                      NaN                            NaN   \n",
       "76                                      NaN                            NaN   \n",
       "77                                      NaN                            NaN   \n",
       "78                                      0.0                       1.111111   \n",
       "79                                      NaN                            NaN   \n",
       "80                                      NaN                            NaN   \n",
       "81                                      NaN                            NaN   \n",
       "82                                      NaN                            NaN   \n",
       "83                                      NaN                            NaN   \n",
       "84                                      0.0                       1.111111   \n",
       "85                                      NaN                            NaN   \n",
       "86                                      NaN                            NaN   \n",
       "87                                      NaN                            NaN   \n",
       "88                                      NaN                            NaN   \n",
       "89                                      NaN                            NaN   \n",
       "90                                      0.0                       1.111111   \n",
       "91                                      NaN                            NaN   \n",
       "92                                      NaN                            NaN   \n",
       "93                                      NaN                            NaN   \n",
       "94                                      NaN                            NaN   \n",
       "95                                      NaN                            NaN   \n",
       "\n",
       "   2b_2b_cut1_total_reads_percent target_plus_plasmid_total_reads_percent  \\\n",
       "0                             0.0                                     0.0   \n",
       "1                             NaN                                     NaN   \n",
       "2                             NaN                                     NaN   \n",
       "3                             NaN                                     NaN   \n",
       "4                             NaN                                     NaN   \n",
       "5                             NaN                                     NaN   \n",
       "6                             0.0                                     0.0   \n",
       "7                             NaN                                     NaN   \n",
       "8                             NaN                                     NaN   \n",
       "9                             NaN                                     NaN   \n",
       "10                            NaN                                     NaN   \n",
       "11                            NaN                                     NaN   \n",
       "12                            0.0                                     0.0   \n",
       "13                            NaN                                     NaN   \n",
       "14                            NaN                                     NaN   \n",
       "15                            NaN                                     NaN   \n",
       "16                            NaN                                     NaN   \n",
       "17                            NaN                                     NaN   \n",
       "18                            0.0                                     0.0   \n",
       "19                            NaN                                     NaN   \n",
       "20                            NaN                                     NaN   \n",
       "21                            NaN                                     NaN   \n",
       "22                            NaN                                     NaN   \n",
       "23                            NaN                                     NaN   \n",
       "24                            0.0                                     0.0   \n",
       "25                            NaN                                     NaN   \n",
       "26                            NaN                                     NaN   \n",
       "27                            NaN                                     NaN   \n",
       "28                            NaN                                     NaN   \n",
       "29                            NaN                                     NaN   \n",
       "..                            ...                                     ...   \n",
       "66                            0.0                                     0.0   \n",
       "67                            NaN                                     NaN   \n",
       "68                            NaN                                     NaN   \n",
       "69                            NaN                                     NaN   \n",
       "70                            NaN                                     NaN   \n",
       "71                            NaN                                     NaN   \n",
       "72                            0.0                                     0.0   \n",
       "73                            NaN                                     NaN   \n",
       "74                            NaN                                     NaN   \n",
       "75                            NaN                                     NaN   \n",
       "76                            NaN                                     NaN   \n",
       "77                            NaN                                     NaN   \n",
       "78                            0.0                                     0.0   \n",
       "79                            NaN                                     NaN   \n",
       "80                            NaN                                     NaN   \n",
       "81                            NaN                                     NaN   \n",
       "82                            NaN                                     NaN   \n",
       "83                            NaN                                     NaN   \n",
       "84                            0.0                                     0.0   \n",
       "85                            NaN                                     NaN   \n",
       "86                            NaN                                     NaN   \n",
       "87                            NaN                                     NaN   \n",
       "88                            NaN                                     NaN   \n",
       "89                            NaN                                     NaN   \n",
       "90                            0.0                                     0.0   \n",
       "91                            NaN                                     NaN   \n",
       "92                            NaN                                     NaN   \n",
       "93                            NaN                                     NaN   \n",
       "94                            NaN                                     NaN   \n",
       "95                            NaN                                     NaN   \n",
       "\n",
       "   plasmid_only_total_reads_percent total_aligned_junctions_collapsed  \\\n",
       "0                               0.0                              68.0   \n",
       "1                               NaN                               NaN   \n",
       "2                               NaN                               NaN   \n",
       "3                               NaN                               NaN   \n",
       "4                               NaN                               NaN   \n",
       "5                               NaN                               NaN   \n",
       "6                               0.0                              68.0   \n",
       "7                               NaN                               NaN   \n",
       "8                               NaN                               NaN   \n",
       "9                               NaN                               NaN   \n",
       "10                              NaN                               NaN   \n",
       "11                              NaN                               NaN   \n",
       "12                              0.0                              68.0   \n",
       "13                              NaN                               NaN   \n",
       "14                              NaN                               NaN   \n",
       "15                              NaN                               NaN   \n",
       "16                              NaN                               NaN   \n",
       "17                              NaN                               NaN   \n",
       "18                              0.0                              68.0   \n",
       "19                              NaN                               NaN   \n",
       "20                              NaN                               NaN   \n",
       "21                              NaN                               NaN   \n",
       "22                              NaN                               NaN   \n",
       "23                              NaN                               NaN   \n",
       "24                              0.0                              68.0   \n",
       "25                              NaN                               NaN   \n",
       "26                              NaN                               NaN   \n",
       "27                              NaN                               NaN   \n",
       "28                              NaN                               NaN   \n",
       "29                              NaN                               NaN   \n",
       "..                              ...                               ...   \n",
       "66                              0.0                              68.0   \n",
       "67                              NaN                               NaN   \n",
       "68                              NaN                               NaN   \n",
       "69                              NaN                               NaN   \n",
       "70                              NaN                               NaN   \n",
       "71                              NaN                               NaN   \n",
       "72                              0.0                              68.0   \n",
       "73                              NaN                               NaN   \n",
       "74                              NaN                               NaN   \n",
       "75                              NaN                               NaN   \n",
       "76                              NaN                               NaN   \n",
       "77                              NaN                               NaN   \n",
       "78                              0.0                              68.0   \n",
       "79                              NaN                               NaN   \n",
       "80                              NaN                               NaN   \n",
       "81                              NaN                               NaN   \n",
       "82                              NaN                               NaN   \n",
       "83                              NaN                               NaN   \n",
       "84                              0.0                              68.0   \n",
       "85                              NaN                               NaN   \n",
       "86                              NaN                               NaN   \n",
       "87                              NaN                               NaN   \n",
       "88                              NaN                               NaN   \n",
       "89                              NaN                               NaN   \n",
       "90                              0.0                              68.0   \n",
       "91                              NaN                               NaN   \n",
       "92                              NaN                               NaN   \n",
       "93                              NaN                               NaN   \n",
       "94                              NaN                               NaN   \n",
       "95                              NaN                               NaN   \n",
       "\n",
       "                                                 Type Percent Editing  \n",
       "0               wt_cut1_total_reads_collapsed_percent       17.647059  \n",
       "1               wt_cut1_total_reads_collapsed_percent             NaN  \n",
       "2               wt_cut1_total_reads_collapsed_percent             NaN  \n",
       "3               wt_cut1_total_reads_collapsed_percent             NaN  \n",
       "4               wt_cut1_total_reads_collapsed_percent             NaN  \n",
       "5               wt_cut1_total_reads_collapsed_percent             NaN  \n",
       "6               wt_cut2_total_reads_collapsed_percent        0.000000  \n",
       "7               wt_cut2_total_reads_collapsed_percent             NaN  \n",
       "8               wt_cut2_total_reads_collapsed_percent             NaN  \n",
       "9               wt_cut2_total_reads_collapsed_percent             NaN  \n",
       "10              wt_cut2_total_reads_collapsed_percent             NaN  \n",
       "11              wt_cut2_total_reads_collapsed_percent             NaN  \n",
       "12  large_deletion_cut1_total_reads_collapsed_percent       35.294118  \n",
       "13  large_deletion_cut1_total_reads_collapsed_percent             NaN  \n",
       "14  large_deletion_cut1_total_reads_collapsed_percent             NaN  \n",
       "15  large_deletion_cut1_total_reads_collapsed_percent             NaN  \n",
       "16  large_deletion_cut1_total_reads_collapsed_percent             NaN  \n",
       "17  large_deletion_cut1_total_reads_collapsed_percent             NaN  \n",
       "18  large_inversion_cut1_total_reads_collapsed_per...       16.176471  \n",
       "19  large_inversion_cut1_total_reads_collapsed_per...             NaN  \n",
       "20  large_inversion_cut1_total_reads_collapsed_per...             NaN  \n",
       "21  large_inversion_cut1_total_reads_collapsed_per...             NaN  \n",
       "22  large_inversion_cut1_total_reads_collapsed_per...             NaN  \n",
       "23  large_inversion_cut1_total_reads_collapsed_per...             NaN  \n",
       "24  large_inversion_cut2_total_reads_collapsed_per...        0.000000  \n",
       "25  large_inversion_cut2_total_reads_collapsed_per...             NaN  \n",
       "26  large_inversion_cut2_total_reads_collapsed_per...             NaN  \n",
       "27  large_inversion_cut2_total_reads_collapsed_per...             NaN  \n",
       "28  large_inversion_cut2_total_reads_collapsed_per...             NaN  \n",
       "29  large_inversion_cut2_total_reads_collapsed_per...             NaN  \n",
       "..                                                ...             ...  \n",
       "66  doner_head_head_cut1_total_reads_collapsed_per...        0.000000  \n",
       "67  doner_head_head_cut1_total_reads_collapsed_per...             NaN  \n",
       "68  doner_head_head_cut1_total_reads_collapsed_per...             NaN  \n",
       "69  doner_head_head_cut1_total_reads_collapsed_per...             NaN  \n",
       "70  doner_head_head_cut1_total_reads_collapsed_per...             NaN  \n",
       "71  doner_head_head_cut1_total_reads_collapsed_per...             NaN  \n",
       "72           1a_1a_cut1_total_reads_collapsed_percent        1.470588  \n",
       "73           1a_1a_cut1_total_reads_collapsed_percent             NaN  \n",
       "74           1a_1a_cut1_total_reads_collapsed_percent             NaN  \n",
       "75           1a_1a_cut1_total_reads_collapsed_percent             NaN  \n",
       "76           1a_1a_cut1_total_reads_collapsed_percent             NaN  \n",
       "77           1a_1a_cut1_total_reads_collapsed_percent             NaN  \n",
       "78           2b_2b_cut1_total_reads_collapsed_percent        0.000000  \n",
       "79           2b_2b_cut1_total_reads_collapsed_percent             NaN  \n",
       "80           2b_2b_cut1_total_reads_collapsed_percent             NaN  \n",
       "81           2b_2b_cut1_total_reads_collapsed_percent             NaN  \n",
       "82           2b_2b_cut1_total_reads_collapsed_percent             NaN  \n",
       "83           2b_2b_cut1_total_reads_collapsed_percent             NaN  \n",
       "84  target_plus_plasmid_total_reads_collapsed_percent        0.000000  \n",
       "85  target_plus_plasmid_total_reads_collapsed_percent             NaN  \n",
       "86  target_plus_plasmid_total_reads_collapsed_percent             NaN  \n",
       "87  target_plus_plasmid_total_reads_collapsed_percent             NaN  \n",
       "88  target_plus_plasmid_total_reads_collapsed_percent             NaN  \n",
       "89  target_plus_plasmid_total_reads_collapsed_percent             NaN  \n",
       "90         plasmid_only_total_reads_collapsed_percent        0.000000  \n",
       "91         plasmid_only_total_reads_collapsed_percent             NaN  \n",
       "92         plasmid_only_total_reads_collapsed_percent             NaN  \n",
       "93         plasmid_only_total_reads_collapsed_percent             NaN  \n",
       "94         plasmid_only_total_reads_collapsed_percent             NaN  \n",
       "95         plasmid_only_total_reads_collapsed_percent             NaN  \n",
       "\n",
       "[96 rows x 187 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_alignments_junction = pd.concat([result_amplicon_df, result_plasmid_df], axis=1)\n",
    "\n",
    "sample_info_filename = os.path.join(directory, 'sample_info.csv')\n",
    "\n",
    "experiments = pd.read_csv(sample_info_filename)\n",
    "\n",
    "\n",
    "results_summary = summarize_results(results_alignments_junction)\n",
    "results_summary\n",
    "#this outputs it in a directory above the N501_701 folder as a summary file. Need to label it or it will be overwritten by next tone\n",
    "results_summary_with_experiments = pd.concat([experiments, results_summary], axis=1)\n",
    "results_summary_with_experiments.to_excel(os.path.join(directory, 'results_summary.xlsx'))\n",
    "\n",
    "results_pivot = melt_results(results_summary_with_experiments)\n",
    "results_pivot.to_excel(os.path.join(directory, 'results_summary_pivot.xlsx'))\n",
    "results_pivot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pipeline2\n",
    "#  This is a second pipeline. Different than Uditas but built in a simliar form. \n",
    "## It uses the read distal to the gene specic primer. This read covers the region downstream of the break. After clipping off anything remaing upstream of the break we make a global alignment to look at genomic integrations and translocations. This also quantifies the alignments.\n",
    "\n",
    "\n",
    "### 0. Use correct_prime function in pipeline1 to arrive at reads that were correctly primed.\n",
    "### 1. Trim correctly primed reads to the break site  \n",
    "### 2. Only use a single end of the pair end reads (distal gene specific primer)\n",
    "### 3. Use bowtie2 genome that has the targeting seq in it added\n",
    "### 4. Do global end-to-end alignment\n",
    "### 5. Do a local alignment\n",
    "### 6. process bam file and pull out the high mapQ reads\n",
    "\n",
    "The idea here is to trim the sequence up to the cut on both reads and then do a global alignment end-to-end agasint the human genome.\n",
    "\n",
    "The idea here is to take the files that are 'trimmed' and 'correct priming' checking. Then to do another round of trimming to get down to the cut site. I need to make another folder for 'trimmed_to_cut'\n",
    "\n",
    "For Tn5 setup the 'Read 1' contains the sequence away from the gene specific primer. So it is the read we should concentration on if we don't do paired end.\n",
    "\n",
    "For LAM the 'Read 2' is the the sequence distal to the gene specific primer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "this is for pipeline2 polb fwd (row 0 in sheet)\n",
      "/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/P_Eric4_BCL2Fastq_only/10000Reads\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "NGS_req-ID                                                                     A000H2GWGF\n",
       "name                                                                        Tn5_Pol50_pbF\n",
       "Sample                                                                               Polb\n",
       "description                                   MiniSeq_K562_PolbHIROS_50per_unsorted_polbF\n",
       "Control sample (Y/N)                                                                    N\n",
       "Notes                                                                                 NaN\n",
       "Dilution                                                                              NaN\n",
       "Cell name_type                                                                       K562\n",
       "UMI_Len                                                                        NNNNNNNNNN\n",
       "IndexI7Primer                                                                      prE368\n",
       "I7_Index_ID                                                          P7_N701_SBS12nextera\n",
       "index_I1                                                                             N701\n",
       "barcode_I1                                                                       TAAGGCGA\n",
       "I5_Index_ID                                                              P5_UMI_N50_Tn5-A\n",
       "index_I2                                                                             N501\n",
       "barcode_I2                                                                       GCGATCTA\n",
       "primer1_ForNest                                                                     prE86\n",
       "primer                                                                             prE361\n",
       "genome                                                                               hg38\n",
       "chr                                                                                  chr8\n",
       "start                                                                         4.23497e+07\n",
       "end                                                                           4.23498e+07\n",
       "strand                                                                                  +\n",
       "guide_1                                                                            Polb#3\n",
       "sequence_guide_1                                                     TAATTTTGTGTGGGTCACCC\n",
       "genome_guide_1                                                                       hg38\n",
       "chr_guide_1                                                                          chr8\n",
       "start_guide_1                                                                    42349873\n",
       "end_guide_1                                                                      42349892\n",
       "strand_guide_1                                                                          +\n",
       "guide_2                                                                            Polb#4\n",
       "sequence_guide_2                                                     TGAAACCAGTTTGGTTACCC\n",
       "genome_guide_2                                                                       hg38\n",
       "chr_guide_2                                                                          chr8\n",
       "start_guide_2                                                                    42350280\n",
       "end_guide_2                                                                      42350299\n",
       "strand_guide_2                                                                          +\n",
       "guide_3                                                                               NaN\n",
       "sequence_guide_3                                                                      NaN\n",
       "genome_guide_3                                                                        NaN\n",
       "chr_guide_3                                                                           NaN\n",
       "start_guide_3                                                                         NaN\n",
       "end_guide_3                                                                           NaN\n",
       "strand_guide_3                                                                        NaN\n",
       "genome_plus_targeting                                              hg38_plus_targetvector\n",
       "Replace_Doner_Name                                                               pE038_MC\n",
       "Replace_Donor                           TGACCCACACAAAATTAGTCTTTTAGCAGACTGGTATGTTTCCAAT...\n",
       "Donor (single strand; double strand)                                                  NaN\n",
       "plamid_name                                                            pE049_WO_redundant\n",
       "plasmid_sequence                        ttacataacttacggtaaatggcccgcctggctgaccgcccaacga...\n",
       "genomic_background                                                                    NaN\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#This is for the sequence to the breaksite for the pipeline of genomic alignment \n",
    "# make sure the sequences are capital\n",
    "\n",
    "\n",
    "print('this is for pipeline2 polb fwd (row 0 in sheet)')\n",
    "seq_primer_to_breaksite = 'TGGTTTTATTTCACCCCATGATAGTAATTATATCACTTCTGATCTGTTAAGAATAGACCTTTTAAAAGTATTGGATAACTTAGAGATGAGACATCTTCAGTTACTCTGTTATTCACCTATTACTCCTTAGGTTACTTGTGAATAATTTTGTGTGGGTCA'\n",
    "\n",
    "#print('this is for pipeline2 polb rev (row 1 in sheet)')\n",
    "#seq_primer_to_breaksite = 'ACAAAAGAGGCCAAGCTGGAGCAGGAAATAGATGCACACGGAGGAAATGGTAGTGGAGTTCAGAGGAGGGAGAGGTTCTTTCTGCCTGGG'\n",
    "\n",
    "#this has some seq homology to the genome but should be ok for this pipeline trimming\n",
    "#print('this is for pipeline2 mCherryRev, row2 in sheet')\n",
    "#seq_primer_to_breaksite = 'GTTAGCAGACTTCCTCTGCCCTCAATCTTTAAGAAAAAAAAAAGTCTAACAATGATTTAGGAATGCTTTGAGGACTTAAATGATCTTATTGGAAACATACCAGTCTGCTAAAAGACTAATTTTGTGTGGGTCA'\n",
    "\n",
    "#this is more reliable than the mCherry rev as it has no sequence homology to the genome\n",
    "#print('this is for pipeline2 mCherryFWD, row3 in sheet')\n",
    "#seq_primer_to_breaksite = 'GCAGGAGCTCGTCGACCCATGGGGGCCCGCCCCAACTGGGGTAACCTTTGGGCTCCCCGGGCGCGACTAGTGAATTCAGATCTGATATCTCTAGAAGTCCTGGG'\n",
    "\n",
    "#check amplicon info and directory\n",
    "print(directory)\n",
    "amplicon_info"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Need to make an new bowtie2 index file that includes targeting for alignment agains the whole genome so it is all in one sheet together. \n",
    "\n",
    "### Other option would be to align it to the amplicon, extract unaligned files and then align to the genome but seems cleaner this way. \n",
    "#### Ideally every sequence is unique between the targeting vector and genome.\n",
    "\n",
    "\n",
    "1. Build your fastas of interest and label .fa files.\n",
    "    1. You need fasta of hg38 or reference genome. You can pull this from downloaded bowtie indexed sampels and then use the following command to turn the index into a fasta file: bowtie2-inspect hg38 > hg38.fa   \n",
    "    2. Put all the fasta files in the same folder\n",
    "2. index the files with bowtie\n",
    "    1. use the command bowtie2-build -f pe038_mc.fa,hg38.fa -p hg38_plus_targetvector\n",
    "    2. this has the -p to make it take less ram in my case.\n",
    "    3. In this case it adds the hg38 and the minicircle targeting file together\n",
    "    4. I have a IntelÂ® Coreâ„¢ i7-5500U CPU @ 2.40GHz Ã— 4 with 15.1 GiB ram and it required about 13.8 gigs of ram and 2 hours to do a hg38+small fasta index\n",
    "    5. be sure to pay attention to the name of the new indexed file. \"hg38_plus_targetvector\" in example above. Add it to the sample_info.csv sheet. Under the tab 'genome_plus_targeting'.\n",
    "    6. you can check it indexed correcctly: bowtie2-inspect -s hg38_plus_targetvector\n",
    "    \n",
    "\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes/hg38_plus_targetvector.2bit\n",
      "env: BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes\n"
     ]
    }
   ],
   "source": [
    "# Assign the file_genome_2bit location. This is needed for pulling sequence from the referene genome by location\n",
    "#need a new build\n",
    "assembly_plus_targetvector = amplicon_info['genome_plus_targeting']\n",
    "file_genome_2bit_plus_target = os.path.join('/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes', assembly_plus_targetvector + '.2bit')\n",
    "print(file_genome_2bit_plus_target)\n",
    "\n",
    "#Needs to have a bowtie build in which the targeting vector is included!!\n",
    "\n",
    "%env BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes\n",
    "\n",
    "#check the env variables\n",
    "#%env"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Trim the break up to the cut site\n",
    "\n",
    "trim_fastq_to_break(directory, amplicon_info, seq_primer_to_breaksite)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#this aligns everything after the break to end-to-end \n",
    "align_afterbreak_end_to_end_genome_global(directory, amplicon_info, assembly_plus_targetvector, keep_sam=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#this aligns everything in the local format wehre it can soft clip the ends\n",
    "align_afterbreaks_genome_local(directory, 1, amplicon_info, assembly_plus_targetvector, keep_sam=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
